core/src/main/scala/epollcat/unsafe/KqueueExecutorScheduler.scala

```diff
@@ -0,0 +1,144 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package epollcat.unsafe
+
+import java.util.ArrayDeque
+import scala.collection.mutable.LongMap
+import scala.concurrent.duration._
+import scala.scalanative.libc.errno
+import scala.scalanative.posix.time
+import scala.scalanative.posix.timeOps._
+import scala.scalanative.posix.unistd
+import scala.scalanative.unsafe._
+import scala.scalanative.unsigned._
+import scala.util.control.NonFatal
+
+import KqueueExecutorScheduler._
+import event._
+import eventImplicits._
+
+private[unsafe] final class KqueueExecutorScheduler(
+    private[this] val kqfd: Int,
+    private[this] val maxEvents: Int)
+    extends EventPollingExecutorScheduler {
+
+  private[this] val changes: ArrayDeque[EvAdd] = new ArrayDeque
+  private[this] val callbacks: LongMap[EventNotificationCallback] = new LongMap
+
+  def poll(timeout: Duration): Boolean = {
+    val timeoutIsInfinite = timeout == Duration.Inf
+    val timeoutIsZero = timeout == Duration.Zero
+    val noCallbacks = callbacks.isEmpty
+    val changeCount = changes.size()
+
+    if ((timeoutIsInfinite || timeoutIsZero) && noCallbacks && changeCount == 0)
+      false // nothing to do here
+    else {
+
+      val changelist = stackalloc[kevent64_s](changeCount.toLong)
+      var change = changelist
+      while (!changes.isEmpty()) {
+        val evAdd = changes.poll()
+        change.ident = evAdd.fd.toULong
+        change.filter = evAdd.filter
+        change.flags = (EV_ADD | EV_CLEAR).toUShort
+        change.udata = EventNotificationCallback.toPtr(evAdd.cb)
+        change += 1
+      }
+
+      val timeoutSpec =
+        if (timeoutIsInfinite || timeoutIsZero) null
+        else {
+          val ts = stackalloc[time.timespec]()
+          val sec = timeout.toSeconds
+          ts.tv_sec = sec
+          ts.tv_nsec = (timeout - sec.seconds).toNanos
+          ts
+        }
+
+      val eventlist = stackalloc[kevent64_s](maxEvents.toLong)
+      val flags = (if (timeoutIsZero) KEVENT_FLAG_IMMEDIATE else KEVENT_FLAG_NONE).toUInt
+      val triggeredEvents =
+        kevent64(kqfd, changelist, changeCount, eventlist, maxEvents, flags, timeoutSpec)
+
+      if (triggeredEvents >= 0) {
+        var i = 0
+        var event = eventlist
+        while (i < triggeredEvents) {
+          if ((event.flags.toLong & EV_ERROR) != 0) {
+
+            // TODO it would be interesting to propagate this failure via the callback
+            reportFailure(new RuntimeException(s"kevent64: ${event.data}"))
+
+          } else if (callbacks.contains(event.ident.toLong)) {
```

Comment:
Suspicious behavior here: in the CI logs it seems that a failure here is being reported quite frequently. However, the value of `event.data` is zero. This seems strange because the manpages say that

> with EV_ERROR set in flags and the system error in data.

https://keith.github.io/xcode-man-pages/kqueue.2.html

Reply:
Btw, I do not see these on my local system, which is macOS 11 intel.

Reply:
Wait, I lied, I do see it on my local 😂

Reply:
Ah, I had an off-by-one, which is why `data` was 0.

Reply:
So for the record the error is a "bad file descriptor". What's more interesting is it only seems to happen when running multiple tests; I cannot reproduce it when running a single test in isolation. This suggests some kind of interference perhaps due to file descriptors being re-used?

Reply:
Hard to tell from the audit trail above, sorry. Is this soup enough for me to try on Apple M1?
Not rushing, just polling.

Reply:
Yes, please, I think it's better to try on M1 sooner rather than later :) All the tests should pass, but there will likely be some stack traces logged.

Reply:
OK, I am off to run this PR on M1.  That will keep me out of your hair for awhile.